docs/architecture/adr-013-multisigs.md

```diff
@@ -0,0 +1,148 @@
+# ADR 013: Encoding standard for Multisignatures
+
+## Changelog 
+
+27-07-2018: Update draft to use amino encoding
+
+11-07-2018: Initial Draft
+
+## Context
+
+Multisignatures, or technically _Accountable Subgroup Multisignatures_ (ASM), 
+are signature schemes which enable any subgroup of a set of signers to sign any message, 
+and reveal to the verifier exactly who the signers were. 
+This allows for complex conditionals of when to validate a signature.
+
+Suppose the set of signers is of size _n_.
+If we validate a signature if any subgroup of size _k_ signs a message,
+this becomes what is commonly reffered to as a _k of n multisig_ in Bitcoin.
+
+This ADR specifies the encoding standard for general accountable subgroup multisignatures,
+k of n accountable subgroup multisignatures, and its weighted variant.
+
+In the future, we can also allow for more complex conditionals on the accountable subgroup. 
+
+## Proposed Solution
+
+### New structs
+
+Every ASM will then have its own struct, implementing the crypto.pubkey interface.
+
+This ADR assumes that [replacing crypto.Signature with []bytes](https://github.com/tendermint/tendermint/issues/1957) has been accepted.
+
+#### K of N threshold signature
+
+The pubkey is the following struct:
+
+```golang
+type ThresholdMultisignatureKey struct { // K of N threshold multisig
+	K       uint               `json:"threshold"`
+	Pubkeys []crypto.Pubkey    `json:"pubkeys"`
+}
+```
+We will derive N from the length of pubkeys. (For spatial efficiency in encoding)
+
+It will expect to verify an `[]byte` encoded version of the Multisignature.
+(Multisignature is described in the next section)
+The multisignature will be rejected if the bitmap has less than k indices,
+or if any signature at any of the k indices is not a valid signature from
+the kth public key on the message.
+(If more than k signatures are included, all must be valid)
+
+The reason this doesn't use `log_8(n)` bytes per signer is because that heavily optimizes for the case where a very small number of signers are required.
+e.g. for `n` of size `24`, that would only be more space efficient for `k < 3`. 
+This seems less likely, and that it should not be the case optimized for.
+
+#### Weighted threshold signature
+
+The pubkey is the following struct:
+
+```golang
+type WeightedThresholdMultisignatureKey struct {
+	Weights []uint             `json:"weights"`
+	Threshold uint             `json:"threshold"`
+	Pubkeys []crypto.Pubkey    `json:"pubkeys"`
```

Comment:
Not sure where these `WeightedThresholdMultisignatureKey` will be used or why they are needed, but I guess this has sth to do with staking? 

I'm wondering if we shouldn't have a type `WeightedPubkeys` instead of two loosely coupled arrays? "Weights and Pubkeys must be of the same length." would be autom. enforced by that. What do you think?

Reply:
The reason I proposed it like this is that it saves space in encoding. Using a WeightedPubkey struct would then have to encode both of the field numbers for each pubkey.

Weighted Threshold Multisignatures are intended for use in the cosmos hub.

Reply:
What does `both of the field numbers` mean here?

Reply:
Every time you encoded a Pubkey then, the amino encoded version would have two field numbers, one for the weight, one for the actual pubkey. 

Having a single "Weights" slice and "pubkeys" slice allows saving the field number from being encoded for each weight/pubkey pair.